In [1]:
import pandas as pd 
# pandas are used for data manipulation, lists and pandas are the only way to manipulate data around.

---
---
---
# Core components, Series and DataFrame
A Series is a column, and a DataFrame is a multi-dimensional table made up of a collection of Series.
![picture](https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png)


---
---
---
# Creating, loading, storing data

---

## Creating DataFrames from Lists

To create a `DataFrame` from lists, we use `pd.DataFrame()` function.

This function will take parameters 
*  `df_1, df_2, df_3` - list we want use as the data source. It has to be in a nested list format - [ [x1, y1], [x2, y2], [x3, y3]  ] into `df` variable

*   `column_names` - names of the columns [ name_x, name_y ]

In [ ]:
df_1 = ["bilu", 19]
df_2 = ["shamuritooo", 21]
df_3 = ["dannu", 16]

df = [df_1, df_2, df_3] 
column_names = ['Name', 'Age']

df = pd.DataFrame(data = df, 
                  columns = column_names) 
df 

Name  Age
0         bilu   19
1  shamuritooo   21
2        dannu   16

Notice that the approach above, requires a new list for every single data row. This is useful when you are adding new rows to an already existing list, like during a `for loop`.

To append a row to `DataFrame` we need to create a new `DataFrame` with the data we wish to append, and then use `df.append()`

In [ ]:
new_data = ["new data", 13]
new_data = [new_data]
df_to_append = pd.DataFrame(data = new_data, 
                            columns = column_names) 

df = df.append(df_to_append)
df

Name  Age
0         bilu   19
1  shamuritooo   21
2        dannu   16
0     new data   13
0     new data   13
0     new data   13

If you have a `list` of values, that represent one variable, and you need to to add it to the `DataFrame` as one column, use `list(zip())` function:


In [ ]:
df_1 = [0, 1, 6, 3, 6]
df_2 = [ 7, 2, 1, 7, 6]

df = list(zip(df_1, df_2))
column_names = ['apples', 'oranges']
index_values = ['bilu', 'shamurito', "Dannu", "rasto", "mato"]

df = pd.DataFrame(data = df, 
                  columns = column_names, 
                  index = index_values)
df

apples  oranges
bilu            0        7
shamurito       1        2
Dannu           6        1
rasto           3        7
mato            6        6

If we wish to append a `row` we use the same approach as before.

In [ ]:
new_data = [4, 4]
df_to_append = pd.DataFrame(data = [new_data], 
                            columns = column_names,
                            index = ["apppend_single"]) 

df = df.append(df_to_append)
df

apples  oranges
bilu                 0        7
shamurito            1        2
Dannu                6        1
rasto                3        7
mato                 6        6
apppend_single       4        4

If you wish to append list with multiple valuies, similarly to before, use `list(zip())` approach to create `DataFrame`, and then use `df.append()` function.

In [ ]:
df_1 = [10, 10, 10]
df_2 = [10, 10, 10]

data_to_append = list(zip(df_1, df_2))
column_names = ['apples', 'oranges']
index_values = ["append_list_zip_1", "append_list_zip_2", "append_list_zip_3"]

df_to_append = pd.DataFrame(data = data_to_append, 
                            columns = column_names, 
                            index = index_values)
df = df.append(df_to_append)
df

apples  oranges
bilu                    0        7
shamurito               1        2
Dannu                   6        1
rasto                   3        7
mato                    6        6
apppend_single          4        4
append_list_zip_1      10       10
append_list_zip_2      10       10
append_list_zip_3      10       10

---

## Reading data from CSV

Read the data a CSV file, and inport it as `DataFrame`, using `pd.read_csv()`

### From CSV file

If the file is saved locally, you can read the file by specifying its location in the hardrive.

To do this in the `Google Colab`, go to the sidebar on the top left of your screen , open the `Files` section, and locate the file you wish to read. `Right click` and `copy path`. The location to the file is now saved, now, `paste` it into the `pd.read_csv()` funtion.

If you want to set a specific column from your dataset as the index, you use `index_col=0` parameter

In [ ]:
df = pd.read_csv("/content/sample_data/california_housing_test.csv")
df

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
1       -118.30     34.26  ...         3.5990            176500.0
2       -117.81     33.78  ...         5.7934            270500.0
3       -118.36     33.82  ...         6.1359            330000.0
4       -119.67     36.33  ...         2.9375             81700.0
...         ...       ...  ...            ...                 ...
2995    -119.86     34.42  ...         1.1790            225000.0
2996    -118.14     34.06  ...         3.3906            237200.0
2997    -119.70     36.30  ...         2.2895             62000.0
2998    -117.12     34.10  ...         3.2708            162500.0
2999    -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]

#### Manually upload files



Manually by dragging and releasing the file onto the side-bar, when you have `Files` section opened. This way you can eaasily create new folders, and sort it out.



---

#### Link Google Drive 

Sign in to your `Google Drive`, and have instantaneous access to files stored there. 

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


Now, when you notice in the `Files` section, you have a folder `gdrive`, where you can locate the data you have stored there, and can use it accordingly. Likewise, you will be able to save graphs, results, anything right onto your `Google Drive`.

#### Link to downloadable CSV file

If you have downloadable link of your data-set, you can use it in the following way.

Note that the link nor the file can be protected by a password, otherwise it will cause problems. 

In [ ]:
url="https://raw.githubusercontent.com/cs109/2014_data/master/countries.csv"
df = pd.read_csv(url)
df

Country         Region
0      Algeria         AFRICA
1       Angola         AFRICA
2        Benin         AFRICA
3     Botswana         AFRICA
4      Burkina         AFRICA
..         ...            ...
189   Paraguay  SOUTH AMERICA
190       Peru  SOUTH AMERICA
191   Suriname  SOUTH AMERICA
192    Uruguay  SOUTH AMERICA
193  Venezuela  SOUTH AMERICA

[194 rows x 2 columns]

---

## Saving DataFrame to CSV

If you wish to save a `DataFrame` as a CSV, you use `nameofyourdataframe.to_csv()` and inside the function you name it `name.csv`, with the extension you want the result to have `.csv` in the example.

In [ ]:
df.to_csv('name_of_the_saved_csv.csv')

You can now go to `Files` section and manually download it.

---

Or, you can set up a code that copies the result to your google drive, by using command:

> `!cp -r name_of_the_saved_file path_to_save_folder_at_google_drive`



Following the previous example, `name_of_the_saved_file` would be `new_purchases.csv`

`path_to_save_folder_at_google_drive` can be obtained by locating it in the `Folders` section `gdrive - My Drive - ...` and using `Copy path` at the destination folder. Note that when you paste it, you will need to use `\` in fron of spaces, otherwise it will give you `cp: target 'something is not a directory`, example:

Copy path result: `/content/gdrive/My Drive/Colab Notebooks/Data`

Change it to this: `/content/gdrive/My\ Drive/Colab\ Notebooks/Data`

In [ ]:
!cp -r new_purchases.csv /content/gdrive/My\ Drive/Colab\ Notebooks/Data

---
---
---

# Important functionality

---

## Load CSV data from a link

Using `pd.read_csv()` with url link to the download destination of the data-set.

In [2]:
url = "https://raw.githubusercontent.com/LearnDataSci/articles/master/Python%20Pandas%20Tutorial%20A%20Complete%20Introduction%20for%20Beginners/IMDB-Movie-Data.csv"
df_movies = pd.read_csv(url)
df_movies.head() # head is first five rows in the dataframe

Rank                    Title  ... Revenue (Millions) Metascore
0     1  Guardians of the Galaxy  ...             333.13      76.0
1     2               Prometheus  ...             126.46      65.0
2     3                    Split  ...             138.12      62.0
3     4                     Sing  ...             270.32      59.0
4     5            Suicide Squad  ...             325.02      40.0

[5 rows x 12 columns]

We want the `Title` column to be the `index`, instead of numbers, so that we can easily search for the entries. To do this, use `index_col="Title"` as a parameter in `pd.read_csv()` function:

In [3]:
df_movies = pd.read_csv(url, index_col="Title")
df_movies.head()

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0

[5 rows x 11 columns]

---

## Viewing data

Using `.head()`, `.tail`, `.sample()`

`.head()` outputs the first five rows of your DataFrame

`.tail()` outputs the last five rows of your DataFrame

`.sample()` outputs the **random** five rows of your DataFrame

All 3 of the above can take a `number as an input`, changing the number of displayed entries from default five to the `specified number`.

In [ ]:
df_movies.head()

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0

[5 rows x 11 columns]

First thing when we load a dataset is always viewing a sample of rows to see how does it look. Here we can see the names of each column, the index, and examples of values in each row.

Personally I prefer to use `.sample()` because it yields random entries. Often what happens is that the data-set is neatly ordered by similar rows, meaning when we use `.head()` it will show entries that are very similar. Using `.sample()` gives a better view of the data-set. 

In [ ]:
df_movies.sample(8)

Rank  ... Metascore
Title                      ...          
Slumdog Millionaire   366  ...      86.0
Law Abiding Citizen   499  ...      34.0
Inland Empire         966  ...       NaN
Free State of Jones   618  ...       NaN
A Good Year           531  ...      47.0
Fast Five             320  ...      66.0
Raw (II)              207  ...      81.0
The Whole Truth       629  ...       NaN

[8 rows x 11 columns]

---

## Getting more info about the data

`.info()` should be one of the very first functions you use after loading your data

It shows essential details, such as number of rows and columns, the number of non-null values, and less imporantly what type of data is in each column (this is only imporant to quickly see if numbers were not accidentally imported as strings).

In [ ]:
df_movies.info() # int is whole number, float is decimal number, object is the rest

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 133.8+ KB


Notice that `Revenue` and `Metascore` have missing values.

Another way to obtain number of rows and columns is `.shape` - without `()`. Just remmber it with `()` and when you run it, it will give you error, and that way you remember () are not supposed to be there.

`.shape` will be used often when making models, and reformating the data.

In [ ]:
df_movies.shape

(1000, 11)

---

## Removing duplicates

It is very imporant to always run this function at the begining to remove any duplicate valuess. Note that some data-sets do contain duplicates as legitimate values, it is important to think if for the current data-set, they have to be removed.

To do this, we use `df.drop_duplicates()` function.

In [ ]:
df_movies = df_movies.drop_duplicates()

Notice that we have say `df_movies = `. This is because using `.drop_duplicates()` creates a copy of the **DataFrame** it is used on. We can either create a new **DataFrame** by calling the variable something else, or we can use the same name to adjust the original.

Alternatively, we can use `inplace=True` parameter, to adjust the original **DataFrame** right away:

In [ ]:
df_movies.drop_duplicates(inplace=True)

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0
...                       ...  ...       ...
Secret in Their Eyes      996  ...      45.0
Hostel: Part II           997  ...      46.0
Step Up 2: The Streets    998  ...      50.0
Search Party              999  ...      22.0
Nine Lives               1000  ...      11.0

[1000 rows x 11 columns]

---

## Cleaning column names

Often original column names contain symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we clean it up at the begining.

To print **DataFrame**'s columns, use `df.columns`

In [ ]:
df_movies.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Now we can quickly read all the columns, and copy-paste it if we try to rename change something. It is also useful if you get `Key Error` when trying to select column by its name... which should never happen, because you will learn to clean columns.

Let's rename **Runtime (Minutes)** and **Revenue (Millions)** to something simpler, using `df.rename()` function:

In [4]:
df_movies.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue millions'
    }, inplace=True)
df_movies.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue millions', 'Metascore'],
      dtype='object')

Note that we used `inplace = True`, which saves us from using `df_movies = `

Now let's change all letters to a lowercase. The most simplest way is to copy-paste the above obtained list of colum names, and manually change everything to lowercase.

In [5]:
df_movies.columns = ['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 
                     'rating', 'votes', 'revenue millions', 'metascore']
df_movies.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue millions', 'metascore'],
      dtype='object')

Alternatively, if there are too many upper-case letters, or too many columns we can use `.lower()` python function, and use a `for loop`.

In [6]:
i = 0
for col in df_movies.columns:
  df_movies.columns.values[i] = col.lower().replace(" ", "_")
  i = i + 1

# or more advanced version:
# df_movies.columns = [col.lower() for col in df_movies]
df_movies.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

You should always lowercase, remove special characters, and replace spaces with underscores if you'll be working with a dataset for some time.

---

## Missing values

Also known as non-existent values, often marked as `null` or `None` or `np.nan` or empty space.

3 things can be done:


1.   Remove columns with missing values
2.   Remove rows with missing values
3.   Replace missing values with actual values, technique known as `inputation`



First, let's check the scale of the missing values. Remember we have already use `df.info()` to get some idea, but let's take it a step further by using `df.isnull()`

In [ ]:
df_movies.isnull().sum()

rank                  0
title                 0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

### Removing columns

If the column contains majority of the missing values (depending on the data-set and problem) but let's say `null > 60%`, then you can consider deleting the column.

If the column values have no meaning, or have a singular value, consider removing the column too.

In [ ]:
df_1 = [0, 1, 6, 3, 6]
df_2 = [ 7, 2, 1, 7, 6]

df = list(zip(df_1, df_2))
column_names = ['apples', 'oranges']
index_values = ['bilu', 'shamurito', "Dannu", "rasto", "mato"]

df = pd.DataFrame(data = df, 
                  columns = column_names, 
                  index = index_values)
df

apples  oranges
bilu            0        7
shamurito       1        2
Dannu           6        1
rasto           3        7
mato            6        6

To drop the column, use `df.drop()` with `column name`  and `axis = 1` parameters:

In [ ]:
df.drop("oranges", 
         axis = 1)

apples
bilu            0
shamurito       1
Dannu           6
rasto           3
mato            6

Notice that `asix = 1`. If you wish to drop a row, you would use `axis = 0`. 

To explain this, if you remmber `.shape` function on `DataFrame`, we get:

In [ ]:
df.shape

(5, 2)

`5` is number of rows, `2` is number of columns.

`(5,2)` - index `0` contains rows, that is why `axis = 0` affect rows.

Likewise, index `1` contains columns, that is why `axis = 1` affects columns.

### Inputation with mean

The most basic inputation technique is by using mean value of the column. 

More advanced approaches would use selected group of entries to create mean - if an entry with a missing value is a horror genre movie, use only that genre for mean inputation (or use director's list of movies, etc...)

Much more advanced techniques would create a predictive model out of the data, and would accurately predict the missing values.

Let's calculate the mean value of the whole column of `revenue_millions`.

In [ ]:
revenue_mean = df_movies['revenue_millions'].mean()
revenue_mean

82.95637614678897

Using the mean value, we will use `df.fillna()` with the value we want to replace missing valus with - `revenue_mean` parameter.

We can use `inplace=True` or we can use `df = df.drop()` approach, they are the same after all.

In [ ]:
df_movies['revenue_millions'].fillna(revenue_mean, inplace=True)

Notice that we used `df_movies['revenue_millions']` not `df_movies`, this is because we only want inpute missing values in the `revenue_millions` column, not the whole data-set. 

Let's check the dataset to see if missing values for `revenue_millions` are gone.

In [ ]:
df_movies.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

## Understanding variables

`df.describe()` creates a summary of the distribution of continuous variables:

In [ ]:
df_movies.describe()

rank         year  ...  revenue_millions   metascore
count  1000.000000  1000.000000  ...       1000.000000  936.000000
mean    500.500000  2012.783000  ...         82.956376   58.985043
std     288.819436     3.205962  ...         96.412043   17.194757
min       1.000000  2006.000000  ...          0.000000   11.000000
25%     250.750000  2010.000000  ...         17.442500   47.000000
50%     500.500000  2014.000000  ...         60.375000   59.500000
75%     750.250000  2016.000000  ...         99.177500   72.000000
max    1000.000000  2016.000000  ...        936.630000  100.000000

[8 rows x 7 columns]

`df.describe()` can be used on categorical variables too:

In [ ]:
df_movies['genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

This tells us that the genre column has `207 unique values`, the `top value` is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

`df.value_counts()` can tell us the frequency of all values in a column:

In [ ]:
df_movies['genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Comedy,Drama                  27
Action,Adventure,Fantasy      27
Animation,Adventure,Comedy    27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

## Correlation between variables

`df.corr()` generates the relationship between each continuous variable.

When preparing data for the model, it is important to remove highly correlated variables, this is because if correlation is close to 1 or -1, one variable perfectly explains the other, in other words, one is just a simple multiple of the other one, this means one of the variables would not provide any extra information in the model, it would only make it worse because of including more variables. ( we will talk about this in later lectures, and about `curse of dimmensionality`)

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation, -1.0 indicates a perfect negative correlation.

In [ ]:
df_movies.corr()

rank      year  ...  revenue_millions  metascore
rank              1.000000 -0.261605  ...         -0.252996  -0.191869
year             -0.261605  1.000000  ...         -0.117562  -0.079305
runtime          -0.221739 -0.164900  ...          0.247834   0.211978
rating           -0.219555 -0.211219  ...          0.189527   0.631897
votes            -0.283876 -0.411904  ...          0.607941   0.325684
revenue_millions -0.252996 -0.117562  ...          1.000000   0.133328
metascore        -0.191869 -0.079305  ...          0.133328   1.000000

[7 rows x 7 columns]

Looking in the first row, first column we see `rank` has a perfect correlation with itself, which is obvious. 

On the other hand, the correlation between `votes` and `revenue_millions` is 0.6.

## DataFrame selecting, slicing, extracting

Similar to `lists`, this section shows how to select (`slice`) rows, columns, and ranges from both.

### By column

In [10]:
genre_col = df_movies['genre']

type(genre_col)
genre_col

Title
Guardians of the Galaxy     Action,Adventure,Sci-Fi
Prometheus                 Adventure,Mystery,Sci-Fi
Split                               Horror,Thriller
Sing                        Animation,Comedy,Family
Suicide Squad              Action,Adventure,Fantasy
                                     ...           
Secret in Their Eyes            Crime,Drama,Mystery
Hostel: Part II                              Horror
Step Up 2: The Streets          Drama,Music,Romance
Search Party                       Adventure,Comedy
Nine Lives                    Comedy,Family,Fantasy
Name: genre, Length: 1000, dtype: object

In [8]:
genre_col = df_movies[['genre']]

type(genre_col)

pandas.core.frame.DataFrame

If we want to select more columns, just add them to the list.

In [9]:
subset = df_movies[['genre', 'rating']]

subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

If you with to select (locate) columns `by index`, we use `df.iloc[what_rows, what_columns]`.

Now, we use classical `list` selecting. 

* To select all rows, `what_rows` will be `:`. 

*If we want to select all columns until 5th one, we use `:5`

In [ ]:
df_movies.iloc[:, 0:5]

rank  ...                                             actors
Title                          ...                                                   
Guardians of the Galaxy     1  ...  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...
Prometheus                  2  ...  Noomi Rapace, Logan Marshall-Green, Michael Fa...
Split                       3  ...  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...
Sing                        4  ...  Matthew McConaughey,Reese Witherspoon, Seth Ma...
Suicide Squad               5  ...  Will Smith, Jared Leto, Margot Robbie, Viola D...
...                       ...  ...                                                ...
Secret in Their Eyes      996  ...  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...
Hostel: Part II           997  ...  Lauren German, Heather Matarazzo, Bijou Philli...
Step Up 2: The Streets    998  ...  Robert Hoffman, Briana Evigan, Cassie Ventura,...
Search Party              999  ...  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...
Nine Lives               1000  ...  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...

[1000 rows x 5 columns]

### By rows

If you want to locate the row entry by name, use `df.loc[]`

In [ ]:
prom = df_movies.loc["Guardians of the Galaxy"]
prom

rank                                                                1
genre                                         Action,Adventure,Sci-Fi
description         A group of intergalactic criminals are forced ...
director                                                   James Gunn
actors              Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...
year                                                             2014
runtime                                                           121
rating                                                            8.1
votes                                                          757074
revenue_millions                                               333.13
metascore                                                          76
Name: Guardians of the Galaxy, dtype: object

If you want to locate the row entry by index, use `df.iloc[what_row]`. Notice that unlike locating columns by index, we will only use one parameter

In [ ]:
prom = df_movies.iloc[1]
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                              7
votes                                                          485820
revenue_millions                                               126.46
metascore                                                          65
Name: Prometheus, dtype: object

In practice, `df.loc[]` will not be used much, you will be utilising indexing, `df.iloc[]` much more. 

The following are the same:

In [ ]:
movie_subset = df_movies.loc['Guardians of the Galaxy':'Sing']

movie_subset = df_movies.iloc[0:4]

movie_subset

rank  ... metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0

[4 rows x 11 columns]

### Conditional selection

In [11]:
(df_movies['director'] == "Ridley Scott").head()

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
Name: director, dtype: bool

In [ ]:
df_movies[df_movies['director'] == "Ridley Scott"]

rank  ... metascore
Title                         ...          
Prometheus                 2  ...      65.0
The Martian              103  ...      80.0
Robin Hood               388  ...      53.0
American Gangster        471  ...      76.0
Exodus: Gods and Kings   517  ...      52.0
The Counselor            522  ...      48.0
A Good Year              531  ...      47.0
Body of Lies             738  ...      57.0

[8 rows x 11 columns]

In [ ]:
df_movies[df_movies['rating'] >= 8.6].head(3)

rank                    genre  ... revenue_millions metascore
Title                                           ...                           
Interstellar       37   Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55       Action,Crime,Drama  ...           533.32      82.0
Inception          81  Action,Adventure,Sci-Fi  ...           292.57      74.0

[3 rows x 11 columns]

Conditional operators in pands use `|` for "or" and `&` for "and".

In [ ]:
df_movies[(df_movies['director'] == 'Christopher Nolan') | (df_movies['director'] == 'Ridley Scott')].head()


rank                     genre  ... revenue_millions metascore
Title                                            ...                           
Prometheus          2  Adventure,Mystery,Sci-Fi  ...           126.46      65.0
Interstellar       37    Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55        Action,Crime,Drama  ...           533.32      82.0
The Prestige       65      Drama,Mystery,Sci-Fi  ...            53.08      66.0
Inception          81   Action,Adventure,Sci-Fi  ...           292.57      74.0

[5 rows x 11 columns]

In [ ]:
df_movies[df_movies['director'].isin(['Christopher Nolan', 'Ridley Scott'])].head()

rank                     genre  ... revenue_millions metascore
Title                                            ...                           
Prometheus          2  Adventure,Mystery,Sci-Fi  ...           126.46      65.0
Interstellar       37    Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55        Action,Crime,Drama  ...           533.32      82.0
The Prestige       65      Drama,Mystery,Sci-Fi  ...            53.08      66.0
Inception          81   Action,Adventure,Sci-Fi  ...           292.57      74.0

[5 rows x 11 columns]

In [ ]:
df_movies[
    ((df_movies['year'] >= 2005) & (df_movies['year'] <= 2010))
    & (df_movies['rating'] > 8.0)
    & (df_movies['revenue_millions'] < df_movies['revenue_millions'].quantile(0.25))
]

rank               genre  ... revenue_millions metascore
Title                                          ...                           
3 Idiots              431        Comedy,Drama  ...             6.52      67.0
The Lives of Others   477      Drama,Thriller  ...            11.28      89.0
Incendies             714   Drama,Mystery,War  ...             6.86      80.0
Taare Zameen Par      992  Drama,Family,Music  ...             1.20      42.0

[4 rows x 11 columns]

## Applying functions

In [ ]:
def rating_function(x):
    if x >= 8.0:
        return "good"
    else:
        return "bad"

In [ ]:
df_movies["rating_category"] = df_movies["rating"].apply(rating_function)

# movies_df["rating_category"] = movies_df["rating"].apply(lambda x: 'good' if x >= 8.0 else 'bad')

df_movies.head(2)

rank  ... rating_category
Title                          ...                
Guardians of the Galaxy     1  ...            good
Prometheus                  2  ...             bad

[2 rows x 12 columns]